go/libkb/pseudonym.go

```diff
@@ -0,0 +1,262 @@
+	}
+	mh := codec.MsgpackHandle{WriteExt: true}
+	var buf []byte
+	enc := codec.NewEncoderBytes(&buf, &mh)
+	err := enc.Encode(input)
+	if err != nil {
+		return [32]byte{}, err
+	}
+	mac := hmac.New(sha256.New, info.HmacKey[:])
+	mac.Write(buf)
+	hmacArr := mac.Sum(nil)
+	var hmac [32]byte
+	if len(hmacArr) != len(hmac) {
+		return [32]byte{}, fmt.Errorf(
+			"Expected array of size %d, got %d",
+			len(hmac), len(hmacArr))
+	}
+	copy(hmac[:], hmacArr)
+	return hmac, nil
+}
+
+func PostTlfPseudonyms(ctx context.Context, g *GlobalContext, pnymInfos []TlfPseudonymInfo) ([]TlfPseudonym, error) {
+	var pnymReqs []tlfPseudonymReq
+	var pnyms []TlfPseudonym
+	for _, info := range pnymInfos {
+		// Compute the pseudonym
+		pn, err := MakePseudonym(info)
+		if err != nil {
+			return nil, err
+		}
+		pnyms = append(pnyms, pn)
+		pnymReqs = append(pnymReqs, tlfPseudonymReq{
+			Pseudonym: pn.String(),
+			Name:      info.Name,
+			ID:        hex.EncodeToString(info.ID[:]),
+			KeyGen:    int(info.KeyGen),
+			HmacKey:   hex.EncodeToString(info.HmacKey[:]),
+		})
+	}
+
+	payload := make(JSONPayload)
+	payload["tlf_pseudonyms"] = pnymReqs
+
+	_, err := G.API.PostJSON(APIArg{
+		Endpoint:    "kbfs/pseudonym/put",
+		JSONPayload: payload,
+		NeedSession: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return pnyms, nil
+}
+
+// GetTlfPseudonyms fetches info for a list of pseudonyms.
+// The output structs are returned in the order corresponding to the inputs.
+// The top-level error is filled if the entire request fails.
+// The each-struct errors may be filled for per-pseudonym errors.
+func GetTlfPseudonyms(ctx context.Context, g *GlobalContext, pnyms []TlfPseudonym) ([]GetTlfPseudonymEither, error) {
+	var pnymStrings []string
+	for _, x := range pnyms {
+		pnymStrings = append(pnymStrings, x.String())
+	}
+
+	payload := make(JSONPayload)
+	payload["tlf_pseudonyms"] = pnymStrings
+
+	var res getTlfPseudonymsRes
+	err := g.API.PostDecodeJSON(
+		APIArg{
+			Endpoint:    "kbfs/pseudonym/get",
+			NeedSession: true,
+			JSONPayload: payload,
+			NetContext:  ctx,
+		},
+		&res)
+	if err != nil {
+		return nil, err
+	}
+
+	// Validate the response
+	if len(res.TlfPseudonyms) != len(pnyms) {
+		return nil, &PseudonymGetError{fmt.Sprintf("invalid server response for pseudonym get: len %v != %v",
+			len(res.TlfPseudonyms), len(pnyms))}
+	}
+	var resList []GetTlfPseudonymEither
+	for i, e := range res.TlfPseudonyms {
+		resList = append(resList, checkAndConvertTlfPseudonymFromServer(ctx, g, pnyms[i], e))
+	}
+
+	return resList, nil
+}
+
+func checkAndConvertTlfPseudonymFromServer(ctx context.Context, g *GlobalContext, req TlfPseudonym, received getTlfPseudonymRes) GetTlfPseudonymEither {
+	mkErr := func(err error) GetTlfPseudonymEither {
+		return GetTlfPseudonymEither{
+			Info: nil,
+			Err:  err,
+		}
+	}
+
+	x := GetTlfPseudonymEither{}
+
+	if received.Err != nil {
+		// This check is necessary because of https://golang.org/doc/faq#nil_error
```

Comment:
Yes! Wrote more comment. Sneaky. `received.Err` has to be lowly typed (idk good words for this) so that json decoder knows what to do.